src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs

```diff
@@ -51,28 +75,16 @@ public override bool IsWhiteSpace
 
                 using (var writer = new EmptyOrWhiteSpaceWriter())
                 {
-                    foreach (var entry in _buffer)
+                    if (_isSingleContentSet)
                     {
-                        if (entry == null)
-                        {
-                            continue;
-                        }
+                        return IsEmptyOrWhiteSpaceCore(SingleContent, writer);
+                    }
 
-                        var stringValue = entry as string;
-                        if (stringValue != null)
-                        {
-                            if (!string.IsNullOrWhiteSpace(stringValue))
-                            {
-                                return false;
-                            }
-                        }
-                        else
+                    for (var i = 0; i < Buffer.Count; i++)
```

Comment:
Since this change helps only if all tag helpers at one depth in a page add at most one entry to each `TagHelperContent`, should probably also avoid this allocation and thereby apply if all tag helpers at one depth in a page add at most one entry to each `TagHelperContent` _or `Clear()` it_. That is, avoid this allocation by exiting early if `_buffer == null` or doing the `null`-conditional dance.